src/ireiat/data_pipeline/assets/demand/faf5_tons.py

```diff
@@ -11,54 +11,155 @@
+        [FAF_TONS_TARGET_FIELD]
+    ].sum()
+
+    # Pivot by origin/destination to calculate percentages for each mode
+    mode_pivot_pdf = known_agg_pdf.pivot_table(
+        index=["dms_orig", "dms_dest"],
+        columns="dms_mode",
+        values=FAF_TONS_TARGET_FIELD,
+        fill_value=0,
+    )
+
+    # Add total tons and calculate percentages
+    mode_pivot_pdf["total_tons"] = mode_pivot_pdf.sum(axis=1)
+    mode_pivot_pdf["truck_pct"] = mode_pivot_pdf[FAFMode.TRUCK] / mode_pivot_pdf["total_tons"]
+    mode_pivot_pdf["rail_pct"] = mode_pivot_pdf[FAFMode.RAIL] / mode_pivot_pdf["total_tons"]
+    mode_pivot_pdf["water_pct"] = mode_pivot_pdf[FAFMode.WATER] / mode_pivot_pdf["total_tons"]
+
+    # Filter unknown origin/destination pairs that don't exist in known modes
+    valid_orig_dest_pairs = known_agg_pdf[["dms_orig", "dms_dest"]].drop_duplicates()
+
+    # Calculate unique origin-destination pairs in unknown modes before filtering
+    original_unknown_pairs = unknown_modes_pdf[["dms_orig", "dms_dest"]].drop_duplicates()
+
+    # Merge unknown modes with the known mode proportions, filter invalid pairs
+    unknown_modes_pdf = unknown_modes_pdf.merge(
+        mode_pivot_pdf[["truck_pct", "rail_pct", "water_pct"]],
+        on=["dms_orig", "dms_dest"],
+        how="left",
+    )
 
-    # Loop through each mode and generate the corresponding demand DataFrame
-    for mode in FAFDemandByMode:
-        mode_str = mode.name  # e.g., "TRUCK", "RAIL"
+    # Filter unknown origin/destination pairs that don't exist in known modes by merging with valid pairs
+    unknown_modes_pdf = unknown_modes_pdf.merge(
+        valid_orig_dest_pairs, on=["dms_orig", "dms_dest"], how="inner"
+    )
 
-        is_by_mode = faf_containerizable_pdf["dms_mode"] == FAFMode[mode_str]
-        faf_mode_pdf = faf_containerizable_pdf.loc[is_by_mode]
+    # Calculate the dropped pairs
+    remaining_unknown_pairs = unknown_modes_pdf[["dms_orig", "dms_dest"]].drop_duplicates()
+    dropped_pairs_count = len(original_unknown_pairs) - len(remaining_unknown_pairs)
 
-        # Group by origin and destination to calculate total tons
-        total_tons_od_pdf = faf_mode_pdf.groupby(["dms_orig", "dms_dest"], as_index=False)[
-            [FAF_TONS_TARGET_FIELD]
-        ].sum()
+    # Log a warning about the dropped origin-destination pairs
+    if dropped_pairs_count > 0:
+        context.log.warning(
+            f"{dropped_pairs_count} unique dms origin-destination pairs were dropped because they don't exist in the known (truck, rail, water) modes."
+        )
 
-        # Log and store the DataFrame, providing the correct output_name for multi-asset
-        context.log.info(f"{mode_str} mode demand generated with {len(total_tons_od_pdf)} records.")
-        publish_metadata(context, total_tons_od_pdf, output_name=mode.value)
+    # Explode unknown mode rows proportionally into truck, rail, water
```

Comment:
I think this section could bemore concise. Something like

```
results = []
for dms_mode, target_pdf, col_name in zip([FAFMode.TRUCK, ...],[truck_pdf, ...],["truck_pct",...]):
  mode_specific_tons = unknown_modes_pdf[FAF_TONS_TARGET_FIELD]*unknown_modes_pdf[col_name])
  mode_specific_df = pd.concat([unknown_modes[["dms_orig","dms_dest"]],pd.Series(mode_specific_tons,name=FAF_TONS_TARGET_FIELD)])
  combined = pd.concat([truck_pdf, mode_specific_df])
  results.append(combined.groupby(["dms_orig","dms_dest"],as_index=False)[[FAF_TONS_TARGET_FIELD]].sum())

return results
```

Obviously with comments. Could also use a function so we don't repeat ourselves quite so much.

Reply:
Replaced repetitive code with a helper function